### DISCUSSION TITLE
[1주차]_1장_사용자 수에 따른 규모 확장성

### DISCUSSION THREAD

**Discussion by minju412** [11th-SSAFY-19/large-scale-system-design#4] [General]
.

**Comment by hoshogi** [comment]
캐시를 해둔 데이터에 변화가 발생했을 때, 일관성을 위해서 그때그때 마다 캐시도 업데이트를 시켜줘야 하는지 
또는, 캐시된 데이터는 원본 데이터와 차이가 발생할 수 있을 것이라는 것을 가정하고 사용하는 것인지 궁금합니다.

**Reply by galug** [reply]
https://inpa.tistory.com/entry/REDIS-%F0%9F%93%9A-%EC%BA%90%EC%8B%9CCache-%EC%84%A4%EA%B3%84-%EC%A0%84%EB%9E%B5-%EC%A7%80%EC%B9%A8-%EC%B4%9D%EC%A0%95%EB%A6%AC

**Reply by minju412** [reply]
trade off를 고려해 정합성이 떨어져도 되는 부분에만 캐시를 쓰는게 좋다..! -현재-

**Reply by gurwls0122** [reply]
캐시의 데이터의 일관성을 유지하는데 어려움이 있다. 따라서, 캐시를 효과적으로 관리하기 위해 적절한 캐시 업데이트 전략을 사용하는것이 중요하다.

1. Write-through 캐시 : 데이터가 캐시에 쓰일 때 동시에 원본 데이터에도 쓰는 방식. 데이터 일관성을 높여주지만, 쓰기 성능이 저하될 수 있다.
2. Write-back 캐시 : 데이터가 캐시에 먼저 쓰이고, 일정 시간이 지나거나 특정 조건이 충족되었을 때 원본 데이터에 쓰인다. 일관성이 떨어질 수 있다.
3. TTL 설정 : 캐시된 데이터에 유효기간을 설정하여, 일정 시간이 지나면 데이터가 자동으로 무효화되도록 하는 방식. 오래된 데이터를 자동으로 갱신가능..

첨언하자면 현재 말대로 정합성이 떨어져도 되는 부분에 캐시쓰는거 맞는것 같습니다. 또한, 쓰기 요청이 많은 곳에는 지양하고 읽기 요청이 많은 곳에서 지향하는 것이 좋다고 생각합니다.

**Comment by bjho606** [comment]
로드밸런서 vs 프록시 서버

로드밸런서 : '부하 분산'이 주 목적
프록시 서버 : '보안', '캐싱'이 주 목적

**Reply by minju412** [reply]
로드밸런서가 프록시 서버의 일종인 느낌도 있다. 명확히 구분하는 것보다는 '역할'로만 봐도 좋을 듯.

**Comment by galug** [comment]
NoSQL 이 빠르다는 이야기를 했을 때, 범위에 대한 검색이 필요한 경우 RDBMS가 NoSQL보다 항상 느리다고 볼 수 있을까요? 
NoSQL과 RDBMS의 사용에 대한 비교를 좀 더 자세한 관점에서 해보면 좋을 것 같습니다.

**Reply by gurwls0122** [reply]
NoSQL 데이터베이스가 RDBMS 보다 항상 빠르다고 할 수 없다. 애초에 설계 철학과 사용 목적이 다르기 때문에 어느 것이 더 빠른지는 특정 상황과 요구사항에 따라 다르다고 합니다.

NoSQL 사용 이유
1. 유연한 스키마
2. 수평적 확장 : 데이터를 여러 서버에 분산 저장하여 대규모 데이터 처리 가능.
3. 고성능 읽기/쓰기 : 특정 작업에 최적화된 데이터베이스가 많다.

RDBMS 사용 이유
1. 고정된 스키마 : 데이터 무결성과 일관성을 유지하기 위해 사용.
2. ACID : 트랜잭션 처리에 강점이 있어 데이터 일관성이 중요한 어플리케이션에 적합하다.

성능 비교
1. 읽기/쓰기 성능 : NoSQL은 단순한 읽기/쓰기 작업에 뛰어난 성능을 보인다. RDBMS는 무결성이 중요하므로 성능 관점에서는 좋지 않다.
2. 확장성 : NoSQL은 수평적 확장에 용이하고, RDBMS는 수직적 확장을 사용하지만, 최신 RDBMS는 수평적 확장도 지원한다.

🔎 RDBMS 확장
1. 샤딩 : 데이터를 여러 분할된 단위로 각 샤드를 다른 데이터 베이스에 저장함.
2. 레플리카 : 
- 마스터 슬레이브 : 하나의 마스터 서버가 쓰기 작업을 처리하고, 여러 슬레이브 서버가 읽기 작업을 처리함.
이외에 클러스터링, 파티셔닝 등.. 이 있습니다.

정리
애초에 NoSQL과 RDBMS는 특정 사용 사레에 맞게 설계되었기 때문에 성능을 비교할 수 없다. 또한, 사용환경과 요구사항에 따라 달라진다.

**Comment by chwangmin** [comment]
로드 밸런서가 1,2,3 서버로 트래픽을 분산하고 있을 때,
만약에, 1번 서버가 다운된다면, 2,3 번 서버로만 트래픽을 보냅니다.
이 상황에서 어떻게 서버가 다운되어 있는 지 확인하는 방법이 궁금합니다.
그리고 1,2,3 트래픽 분산을 2,3 번 트래픽으로 분산하는 흐름이 궁금합니다.

**Reply by minju412** [reply]
⭐️로드밸런서의 트래픽 분산 알고리즘 익혀두기⭐️

**Reply by gurwls0122** [reply]
## 서버 상태 확인 방법
헬스 체크 : 로드 밸런서는 주기적으로 서버의 상태를 확인하기 위해 헬스 체크를 수행하는데 ping 체크(서버에 ICMP 패킷을 보낸다), TCP 체크(특정 포트에 TCP연결을 시도), HTTP/HTTPS 체크(특정 URL에 HTTP/HTTPS 요청을 보낸다)

## 서버 다운 시 트래픽 분산 흐름
1. 서버 다운 감지 : 헬스 체크를 통해 서버가 응답하지 않거나 오류를 반환하는 것을 감지.
2. 서버 상태 업데이트 : 로드 밸런서가 이를 인지하고 서버 목록에서 다운된 서버를 제거한다.
3. 트래픽 재분산 : 하나의 서버가 제거된 상태에서 다른 서버로 트래픽을 분산한다(트래픽 분산 알고리즘에 따라 이루어짐)
4. 서버 복구 시도 : 주기적으로 1번 서버의 상태를 다시 확인하며 서버가 복구되었는지 확인한다.

## 로드 밸런서의 트래픽 분산 알고리즘
1. 라운드 로빈 : 순차적으로 각 서버에 트래픽을 분산
2. 가중치 라운드 로빈 : 서버에 가중치를 부여하고 가중치에 따라 트래픽을 분산
3. 최소 연결 : 현재 연결된 세션 수가 가장 적은 서버에 트래픽을 분산
4. IP 해시 : ip 주소를 해시 함수에 입력하여 특정 서버에 매핑
5. 가중치 최소 연결 : 가중치를 고려하여 제일 작은 서버에 트래픽을 분산.
6. 응답시간 : 서버 응답시간을 기준으로 가장 빠른 응답을 보이는 서버에 트래픽을 분산.

**Comment by hoshogi** [comment]
![스크린샷 2024-06-27 오후 9 56 51](https://github.com/11th-SSAFY-19/large-scale-system-design/assets/71253970/99684845-9f75-4380-af69-4b4929e468e5)

위 그림과 같이 Stateless 웹 계층 구조에서 세션을 외부 저장소인 NoSQL에 저장한다고 책에 적혀있는데 이때 주로 어떤 NoSQL을 사용해서 세션을 저장하는 지 궁금합니다.

**Reply by gurwls0122** [reply]
NoSQL에서 사용하는 모든 DB마다 장단점이 있다고 합니다(환경/특성에 따라 알아서 잘 쓰라나 뭐라나...)

그래도 두가지 정도 확인해봤는데요.
아래는 아마존에서 두 스토리지에 대해 비교했답니다.
![image](https://github.com/11th-SSAFY-19/large-scale-system-design/assets/80228712/806c0dd1-ffcc-403e-8d9c-6015a2fe06c5)

## Advanced data structures
Redis, Memcached 모두 기본적으로 key-value 방식의 저장소이다.
Memcached는 key-value 둘다 String으로 지원하고,
Redis는 다양한 자료 구조(list, sets, hashes 등..)를 지원한다.

## Multithreaded architecture
Redis는 기본적으로 싱글 스레드로 구현되어있는 반면 Memcached는 멀티 스레드로 처리함.
(단, redis 6.0부터는 부분적으로 multithread를 지원한다고 합니다)

소량의 데이터 즉 100k 이하는 redis를 사용하는 것이, 대용량의 데이터를 처리할 때는 Memcached가 더 높은 성능을 낸다고 합니다.

### Why Memcached가 대용량 처리시 빠름..?
아마 메모리 할당 때문인것 같은데
- redis 는 c언어에서 메모리 할당 및 해제시 malloc, free를 사용하여 할당하게 됨. 즉, 메모리 단편화가 발생하게 되며 파편화로 인해 메모리 할당 시 딜레이가 되어 응답속도에 영향을 준다.
- Memcached는 효율적인 메모리 할당을 위해 Slab이라는 기술을 이용해 파편화를 줄인다.
🔎 slab?
- 메모리를 작은 청크단위로 나누고, 크기를 그룹화하여 메모리 파편화를 최소화하고 성능을 최적화 한다.
출처 : https://studyposting.tistory.com/38

### 하지만 redis..?
In-Memory 기반 DB들은 기본적으로 영속성을 지원하지 않는데
redis는 영속성을 지원하기 위해 Replication을 사용한다(Master의 모든 데이터를 Slave로 복제화하여 이중화를 구성)
Memcached는 Consistent Hashing 개념을 사용하여 장애가 발생한 서버의 데이터만 사라지고 나머지 서버의 데이터들은 사용 가능.

## 결론
- Redis가 더 좋은거 같음!

**Reply by hoshogi** [reply]
https://escapefromcoding.tistory.com/704
Write이외에는 Redis가 Memcached보다 빠르다는 글
